Framework/Geometry/src/Instrument/DetectorInfo.cpp

```diff
@@ -228,6 +228,52 @@ DetectorInfo::signedTwoTheta(const std::pair<size_t, size_t> &index) const {
   return angle;
 }
 
+double DetectorInfo::azimuthal(const size_t index) const {
+  if (isMonitor(index))
+    throw std::logic_error("Azimuthal angle is not defined for monitors");
+
+  const auto samplePos = samplePosition();
+  const auto beamLine = samplePos - sourcePosition();
+
+  if (beamLine.nullVector()) {
+    throw Kernel::Exception::InstrumentDefinitionError(
+        "Source and sample are at same position!");
+  }
+
+  const auto sampleDetVec = position(index) - samplePos;
+  const auto horizontal =
+      m_instrument->getReferenceFrame()->vecPointingHorizontal();
+
+  const double dotHorizontal = sampleDetVec.scalar_prod(horizontal);
```

Comment:
Thanks to the cross product your new vertical is now orthogonal to the beamline. However, the horizontal still is not, so I think you still end up with a discrepancy?

Reply:
I think it would be hard to have happen since `ReferenceFrame` only allows the cardinal directions in its constructor https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/ReferenceFrame.cpp#L22

Reply:
The point is: you can move source and sample, and end up with a beam that is not perpendicular.